alf/networks/s5/s5.py

```diff
@@ -0,0 +1,421 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Callable, Tuple
+import math
+import numpy as np
+import scipy
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+import alf
+import alf.layers
+from alf.initializers import variance_scaling_init
+from alf.networks import Network
+
+from .utils import make_DPLR_HiPPO, diag_ssm_forward
+"""
+Implement the S5 networks described in
+
+J. T.H. Smith et. al. Simplified State Space Layers for Sequence Modeling
+https://openreview.net/forum?id=Ai8Hw3AXqks
+
+The original JAX implementation is at https://github.com/lindermanlab/S5
+
+
+The following modules are implemented:
+
+- S5SSM: corresponds to s5.ssm.S5SSM. The core diagonal state space model
+- S5Block: corresponds to s5.layers.SequenceLayer. A
+- create_stacked_s5_encoder: corresponds to s5.seq_model.StackedEncoderModel
+
+An example of sMNIST training is in s5_train.py
+"""
+
+
+def _view_as_complex(x: torch.Tensor) -> torch.Tensor:
+    """
+    Args:
+        x: real tensor with shape [..., N]
+    Returns:
+        complex tensor with shape [..., N // 2]
+    """
+    return torch.view_as_complex(x.reshape(*x.shape[:-1], x.shape[-1] // 2, 2))
+
+
+def _view_as_real(x):
+    """
+    Args:
+        x: complex tensor with shape [..., N]
+    Returns:
+        real tensor with shape [..., 2*N]
+    """
+    return torch.view_as_real(x).reshape(*x.shape[:-1], x.shape[-1] * 2)
+
+
+def _as_real_to_complex_matrix(x: torch.Tensor) -> torch.Tensor:
+    """
+
+    .. code-block:: python
+
+        B = _as_real_to_complex(A)
+        assert (B @ x == _view_as_real(A @ x.to(B.dtype))).all()
+
+    Args:
+        x (torch.Tensor): complex matrix with shape [M, N]
+    Returns:
+        torch.Tensor: real matrix with shape [2*M, N]
+
+    """
+    return torch.view_as_real(x).mT.reshape(2 * x.shape[0], x.shape[1])
+
+
+def _as_complex_to_real_matrix(x: torch.Tensor) -> torch.Tensor:
+    """
+
+    .. code-block:: python
+
+        B = _as_complex_to_real_matrix(A)
+        assert (B @ _view_as_real(x) == (A @ x).real).all()
+
+    Args:
+        x: complex matrix with shape [M, N]
+    Returns:
+        real matrix with shape [M, 2*N]
+    """
+    return _view_as_real(x.conj().resolve_conj())
+
+
+class S5SSM(Network):
+    r"""SSM with diagonal state transition matrix.
+
+    Corresponds to s5.ssm.S5SSM
+
+    The model is defined as the following in continuous time:
+
+    .. math::
+
+        ds/dt = As + Bu
+        A = V \Lambda V^{-1}
+        ds/dt = V \Lambda V^{-1}s + Bu
+        dV^{-1}s/dt = \Lambda V^{-1}s + V^{-1}Bu
+        y = C x + D u
+        y = CV V^{-1}s + Du
+
+    where :math:`s` is the state, :math:`u` is the input, :math:`y` is the output,
+    :math:`A` is the state transition matrix, :math:`B` is the input matrix,
+    :math:`C` is the output matrix, :math:`D` is the feedforward matrix,
```

Comment:
Maybe mention `D` her is also assumed to be diagonal?

Reply:
added